app/src/androidTest/java/com/github/h3lp3rs/h3lp/forum/ForumCacheTest.kt

```diff
@@ -0,0 +1,166 @@
+package com.github.h3lp3rs.h3lp.forum
+
+import androidx.test.core.app.ApplicationProvider
+import com.github.h3lp3rs.h3lp.database.Databases.*
+import com.github.h3lp3rs.h3lp.database.Databases.Companion.setDatabase
+import com.github.h3lp3rs.h3lp.database.MockDatabase
+import com.github.h3lp3rs.h3lp.forum.implementation.CachedForum
+import com.github.h3lp3rs.h3lp.signin.SignInActivity.Companion.globalContext
+import com.github.h3lp3rs.h3lp.storage.Storages.Companion.resetStorage
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+
+class CacheForumTest {
+
+    // Useful variables
+    private lateinit var rawForum: Forum
+    private lateinit var cachedForum: Forum
+
+    @Before
+    fun setup() {
+        globalContext = ApplicationProvider.getApplicationContext()
+        resetStorage()
+        setDatabase(PREFERENCES, MockDatabase())
+        ForumCategory.mockForum()
+        rawForum = ForumCategory.root().child(ForumCategory.GENERAL.name)
+        cachedForum = CachedForum(rawForum)
+    }
+
+    private fun resetOnlineForum() {
+        ForumCategory.mockForum()
+        rawForum = ForumCategory.root().child(ForumCategory.GENERAL.name)
+        cachedForum = CachedForum(rawForum)
+    }
+
+    @Test
+    fun newPostIsAddedToCache() {
+        // Add a post
+        val postOnline = cachedForum.newPost(AUTHOR, CONTENT, isPost = true).join()
+
+        // Reset the forum, acts as a clear of the data
+        resetOnlineForum()
+
+        // Fetch through cache the post
+        val postCached = cachedForum.getPost(postOnline.post.key).join()
+
+        assertEquals(CONTENT, postCached.post.content)
+        assertEquals(AUTHOR, postCached.post.author)
+    }
+
+    @Test
+    fun repliesAreAddedToCache() {
+        // Add a post
+        val postOnline = cachedForum.newPost(AUTHOR, CONTENT, isPost = true).join()
+
+        // Add two replies
+        postOnline.reply(AUTHOR, CONTENT).join()
+        postOnline.reply(AUTHOR, CONTENT).join()
+
+        // Reset the forum, acts as a clear of the data
+        resetOnlineForum()
+
+        // Fetch through cache the post
+        val postCached = cachedForum.getPost(postOnline.post.key).join()
+
+        assertEquals(2, postCached.replies.size)
+    }
+
+    @Test
+    fun postAndGetDoesNotDuplicate() {
+        // Add a post
+        val postOnline = cachedForum.newPost(AUTHOR, CONTENT, isPost = true).join()
+
+        // Fetch it again
+        cachedForum.getPost(postOnline.post.key).join()
+
+        // Reset the forum, acts as a clear of the data
+        resetOnlineForum()
+
+        // Fetch through cache the post
+        val all = cachedForum.getAll().join()

```

Comment:
This only seems to work as if the online forum's getAll returns an empty list, you take the value from getAllAux, but if the forum's getAll had returned exceptionally, I think this wouldn't work (c.f. my comment on getAll in the CachedForum)

Reply:
No, as already explained this is not the case.